### DISCUSSION TITLE
why-occurs-deadlock-from-select-query/

### DISCUSSION THREAD

**Discussion by giscus** [Hyeon9mak/Hyeon9mak.github.io#185] [Blog comments]
# why-occurs-deadlock-from-select-query/

“아니, 도대체 왜 조회 쿼리 메서드에서 데드락이 생기는거야?”  

https://hyeon9mak.github.io/why-occurs-deadlock-from-select-query/

**Comment by JiwonDev** [comment]
현구님 글 정말 잘 쓰시네요 👍 재미있게 읽어서 코멘트 쓱 남겨봅니다.

> 묻는 질의 쿼리를 추가하여, 중복 update 빈도를 줄여 deadlock 이 발생 가능성을 낮춰두었다.

쿼리 후 update 는 동시에 여러 트랜잭션이 A를 읽고 수정하면, 각각 B,C,D로 바꾸더라도 마지막 수정만 반영될 듯 한데요. 그게 큰 상관이 없다면 아예 Entity 조회나 select 를 없애버리고 단건 update 쿼리로 변경하는 건 어떨까요? 

동작은 기존과 동일할테고, contentId로 단건 row만 변경하는거라면 데드락이 발생할 일은 거의 없을 듯 해요. 의견이 궁금합니다 😀

---

> JPA 는 JPQL 을 DB 로 던지기 전 정말로 flush 를 먼저 호출할까?

JPQL 실행 시 정합성을 맞추기 위해 무조건 flush() 실행한다고 많이 설명하는데, 실제로 테스트 해보니 항상 그런건 아니더라구요. 살짝 혼동이 있을 수 있어서 코멘트 남겨둡니다.

정확히는 (flush AUTO) 일 때  쿼리 때문에 영속성 컨텍스트와 DB가 달라질 수 있을 때만 플러시를 합니다. 
- 트랜잭션을 커밋하기 전 / 네이티브 SQL 실행하기 전
- 영속성 컨텍스트의 쓰기 지연 저장소에 **대기 중인 Entity의 작업과 겹치는** JPQL 쿼리를 실행하기 전

예를 들어 한 트랜잭션 내에 
`A 생성 -> JPQL A id 조회 -> A 삭제 -> JPQL A id 다시 조회 -> 커밋`
했을 때, 무조건 flush가 발생한다면 db delete 쿼리와 의미없는 db 조회 쿼리가 찍혀야 겠지만, 실제로 찍어보면 delete 쿼리도 발생하지 않고 마지막 조회에 null이 정상 반환됩니다.
 
이런 부분 때문에 애초에 JPA 사용 시 트랜잭션 내에 조회와 안겹치게, 조회 후 마지막에 수정하는 걸 권장한다고 하네요. 저도 아래 글을 읽고 테스트 해봤었습니다 😀
https://vladmihalcea.com/how-does-the-auto-flush-work-in-jpa-and-hibernate/
https://www.baeldung.com/spring-jpa-flush

직접 실험해볼 때, 아래 로그를 활성화 해보면 JPA 플러시를 자세하게 확인해볼 수 있습니다.
logging.level.org.hibernate.event.internal.AbstractFlushingEventListener: DEBUG

**Reply by Hyeon9mak** [reply]
오 지원님~ 감사합니다 :+1:

> 쿼리 후 update 는 동시에 여러 트랜잭션이 A를 읽고 수정하면, 각각 B,C,D로 바꾸더라도 마지막 수정만 반영될 듯 한데요. 그게 큰 상관이 없다면 아예 Entity 조회나 select 를 없애버리고 단건 update 쿼리로 변경하는 건 어떨까요?

말씀해주신것과 같은 결로, 트랜잭션 자체를 짧게 가져갈 수 있도록 리팩터링하는게
가장 근원적인 해결책이고 좋다고 생각해요. :+1: 
글에 적었던 것처럼 당장 급하게 해결하기 위해 위와 같은 방식을 채택했어요!
더 멋진 구조로 빠르게 변경해보겠습니다 :+1:

> 예를 들어 한 트랜잭션 내에
`A 생성 -> JPQL A id 조회 -> A 삭제 -> JPQL A id 다시 조회 -> 커밋`
했을 때, 무조건 flush가 발생한다면 db delete 쿼리와 의미없는 db 조회 쿼리가 찍혀야 겠지만, 실제로 찍어보면 delete 쿼리도 발생하지 않고 마지막 조회에 null이 정상 반환됩니다.

아하~ 말씀하신 예시는 ID 생성 전략이 `GenerationType.AUTO` 라고 가정하신거군요.

ID 생성 전략이 `GenerationType.AUTO` 인 경우에는 말씀하신 것처럼 동작하나, 
제가 예시로 첨부한 코드와 같이 `GenerationType.IDENTITY` 를 사용하는 경우
ID 생성 전략을 DB 한테 위임해서 DB 에 다녀오기 전까진 ID 를 알수 없으므로 쿼리가 발생해요.

대부분은 flush AUTO 설정을 두고, `GenerationType.IDENTITY` 를 사용하기 때문에
JPQL 수행시에 무조건 쿼리가 발생한다고 이야기하는 것 같아요.
하지만 지원님 설명이 가장 정확한 설명이네요!! :+1:

**Comment by Junuu** [comment]
헐 조회를 하는데 update가 날라가다니.. 
덕분에 배워갑니다ㅎㅎ
알쏭달쏭한 JPA의 세계

**Reply by Hyeon9mak** [reply]
'orm 없이 native query 만 사용하는 환경에서는 이런 고민을 할 필요가 없을텐데...' 라는 생각이 많이 드네요 ㅎㅎ

**Comment by Be-poz** [comment]
간만에 JPA 관련 글 보는 것 같아서 공부하던 시절 기억도 나고 너무 재밌게 봤어

한창 공부하고 사이드프로젝트 할 때에는 entity 값을 수정하고 더티체킹 이용해서 update 날라가게끔 했었는데, 입사하고 나서 부터는 기존 프로젝트 들의 코드도 그렇고 내가 새롭게 짤 때에도 entity 값 수정으로 update를 날리는 방식이 아니라 @Query, queryDSL 을 이용해서 아예 update 용 메서드를 호출하는 방식으로 짜게되는 것 같아. 명시적으로 update가 날라간다라는게 보여서 확 와닿는 느낌(물론 entity 변경을 할 때에도 해당 비즈니스 로직에 대한 메서드 호출이 있긴 하지만) 

한창 jpa 공부할 때에는 orm 너무 좋다 하면서 사용을 했었는데, 정작 지금은 도메인들의 덩치가 커지고 msa 환경이다보니 도메인 엔티티 내부에 다른 엔티티와 연관관계를 맺는 것이 아닌 id 값 자체를 가진 필드가 추가되는 것을 보면 적절하게 섞어 사용하는 것이 역시 맞구나 하고 많이 생각하게 되는 것 같아

**Reply by Hyeon9mak** [reply]
> 명시적으로 update가 날라간다라는게 보여서 확 와닿는 느낌

백번 천번 공감혀~ 특히 위 문장이 와닿네. 👍 

우리가 객체지향 프로그래밍, 클린코드, 클린 아키텍처를 공부하는 이유는 결국 "아무런 배경지식 없이도 이해하기 쉬운 코드를 만들자" 잖아.
그런 면에서 orm 이 제공해주는 편의기능들보다 직관적인 query 가 나은 상황이 많은거 같아.
orm 이 주는 편리함을 누리는 상황 대비 곤란함을 느끼는 상황이 더 많아서 그렇게 생각하는거 같기도 하고 😂

**Comment by LenKIM** [comment]
글을 재미있게 읽어서 여러번 읽었는데, 제가 이해한 내용이 맞는지 궁금해서 다시 여쭤봐요.
- 처음에는 예상으로 동시간대에 하나의 컨텐츠에 여러번의 update 가 발생하는것 아닌가? 라는 가설을 측정했는데, 동료분(경수님)이 로그를 찾아보니 가설대로 동작되는 것을 찾을 수 있었다.

- 그럼 왜 데드락이 걸리는 걸까? Custom 하게 만든 JPQL 의 경우에는 `곧바로` DB Connection 을 가져오고, flush 를 발생시키는 반면에, JPA에서 자동으로 만들어주는 Query Creation 만든 쿼리는 DBConnection,flush 를 발생시키지 않고, 트랜잭션이 끝날때 flush 를 발생시킨다.

- 그래서 다시 왜 데드락이 걸리는 걸까? 원인은 커스텀하게 만든 `findContentByContentId`에 있는데, 커스텀하게 만들었기 때문에 DBConnection, flush 을 빈번하게 발생하게 되는데- findContentByContentId 를 통해 업데이트를 하면서  데드락에 걸리게 되었다.

이렇게 이해했어요~!!

혹시 제가 이해한 내용이 맞을까요?

**Reply by LenKIM** [reply]
단순히 기본적인 update 쿼리가 여러번 일어나서 데드락이 걸린게 아니라, 

custom 한 `findContentByContentId` 조회 후 단순히 기본적인 update 쿼리 동작되는데 `findContentByContentId` 때문에 flush 가 빈번하게 발생해서 데드락이 발생했다는 이야기일까요?

**Reply by Hyeon9mak** [reply]
안녕하세요 정규님! 우선 재밌게 봐주셔서 감사합니다.
글을 두서없이 써놔서 이해가 어려우셨을거 같아요 :scream:

우선 데드락이 발생한 원인은 아주 간단명료합니다.
동시에 2개 이상의 트랜잭션에서 하나의 row 에 update query 를 던졌기 때문에 발생했습니다.
오해가 발생한 지점은 `트랜잭션 2개가 어떻게 만들어졌느냐` 였는데요,

1. 처음에 데드락이 발생한 이유로 예상한 것은 `하나의 메세지(요청) 내부에서 여러 트랜잭션 분기가 일어났고(동시에 2개 이상의 트랜잭션이 생성되었고), 하나의 콘텐츠에 여러 트랜잭션들이 동시에 update query 를 날리다가 데드락이 발생한다.` 였습니다.
2. `하나의 콘텐츠에 여러 트랜잭션들이 동시에 update query 를 날리다가 데드락이 발생한다.` 까진 맞았으나, `하나의 메세지(요청) 내부에서 여러 트랜잭션 분기가 일어났고(동시에 2개 이상의 트랜잭션이 생성되었고)` 가 틀렸었습니다.
3. 동료분께서 시간대별 로그를 확인해보시니 `동시에 2개의 메세지(요청)이 전달` 되고 있었습니다. 그러니 자연스럽게 트랜잭션이 2개 이상이었고, 하나의 콘텐츠에 대해 동시에 접근을 시도해서 경쟁이 발생했어요.

여기까지가 데드락 이야기의 전부입니다.

그럼 조회 쿼리 메서드 이야기는 무엇인가? `데드락이 발생한 정확한 시점`에 대한 이야기였는데요.
분명 데드락은 create / update / delete 쿼리가 DB 로 전달될 때 발생할텐데, read 쿼리가 DB 로 전달되는 시점에서 데드락이 발생한게 궁금해서 파보게 되었습니다. 
짧게 정리해보면... 

1. JPA 는 그 특성상 트랜잭션 중간에는 영속성 컨텍스트에 쿼리들을 모아두었다가 트랜잭션이 종료되면 한방에 던지는데
2. `findContentByContentId` 와 같은 커스텀 쿼리 메서드 때문에 트랜잭션 진행 도중 영속성 컨텍스트에 모여있던 쿼리가 일시적으로 DB 에 전달되었고
3. 그 사이에 들어있던 update query 가 트리거가 되어 해당 시점에 데드락이 발생했다.

... 가 되겠습니다!

**Comment by px201226** [comment]
"하나의 콘텐츠에 대해 동시에 접근을 시도해서 경쟁이 발생했어요."
session1) ....
session2) ...
session1) update content .. where id = 1;
session2) update content .. where id = 1;

이런 상황으로 이해했는데, 이런 조건이라면 데드락이 아니라 타임아웃이 발생해야되지 않나요?
데드락이란게 서로가 원하는 리소스를 참조하려는 상황에 발생하는데, 같은 리소스를 참조한다면 경합은 생기겠지만 데드락이 아니라 한 트랜잭션이 락을 획득하기 위해 기다리거나 타임아웃이 나야될 것 같아서요..

**Reply by Hyeon9mak** [reply]
안녕하세요 기수님! 댓글을 늦게 발견해서 답변이 늦었습니다. :pray:

동일한 행에 대해서만 업데이트를 한다면, 기수님 말씀대로 데드락은 발생하지 않습니다.
대신 위 그림의 예시는 두 트랜잭션이 2개 이상의 row 에 작업을 수행하면서 데드락이 발생했습니다.

- 트랜잭션A 는 1을 업데이트하고, 4를 업데이트 하려고 한다.
- 트랜잭션B 는 2를 업데이트하고, 4를 업데이트 하려고 한다.

"두 트랜잭션이 4 를 기다린다." 에서 데드락이 발생했다고 이해해주시면 될거 같습니다.

**Comment by sju3358** [comment]
글 잘 읽었습니다.
요약하면
그간의 요청들로 인해 영속성이 마구마구 업데이트 되었는데, 
여기서 데드락을 발생시킬 수 있는 Update가 동시에 영속성에 적용되었고,
근데 아직 flush는 안해서 실제로 데드락은 발생을 안했고,

JPQL로 작성한 조회쿼리를 작성시 먼저 Flush를 한번에 호출하게 되고,
이게 트리거가 되서 결국 데드락이 발생한 것인가요?

**Reply by Hyeon9mak** [reply]
안녕하세요 성락님!

> 그간의 요청들로 인해 영속성이 마구마구 업데이트 되었는데,
여기서 데드락을 발생시킬 수 있는 Update가 동시에 영속성에 적용되었고,

요 부분은 제가 잘 이해하지 못했습니다...
영속성 컨텍스트가 변경되는 걸 말씀해주신 걸까요?
추가로 답변 주시면 도움을 드릴 수 있을것 같습니다! 🙇‍♂️ 

> 아직 flush는 안해서 실제로 데드락은 발생을 안했고,
JPQL로 작성한 조회쿼리를 작성시 먼저 Flush를 한번에 호출하게 되고,
이게 트리거가 되서 결국 데드락이 발생한 것인가요?

요 부분은 말씀해주신게 정확합니다! 👍
